### PR TITLE
PYTHON-3004 Support kmip FLE KMS provider

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Resync CSFLE spec tests.
Implement new KMS TLS Options Tests and other kmip prose tests.

### CODE REVIEW

**Comment by ShaneHarvey** [comment]
Waiting for the tests to pass before requesting a review.

**Comment by ShaneHarvey** [comment]
Starting the mock kms server is failing on EVG: 
```
 [2021/11/12 01:05:15.368] + python -u kms_http_server.py --ca_file ../x509gen/ca.pem --cert_file ../x509gen/server.pem --port 8002 --require_client_cert
 [2021/11/12 01:05:15.684] /data/mci/95a1a58fd6dec98b9c985856ff7214a8/src
 [2021/11/12 01:05:15.684] Mock KMS Web Server Listening on port 8000
 [2021/11/12 01:05:15.685] Mock KMS Web Server Listening on port 8002
 [2021/11/12 01:05:15.685] Mock KMS Web Server Listening on port 8001
 [2021/11/12 01:05:16.124] Process SyncManager-1:
 [2021/11/12 01:05:16.124] Traceback (most recent call last):
 [2021/11/12 01:05:16.124]   File "/opt/python/3.6/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
 [2021/11/12 01:05:16.124]     self.run()
 [2021/11/12 01:05:16.124]   File "/opt/python/3.6/lib/python3.6/multiprocessing/process.py", line 93, in run
 [2021/11/12 01:05:16.124]     self._target(*self._args, **self._kwargs)
 [2021/11/12 01:05:16.124]   File "/opt/python/3.6/lib/python3.6/multiprocessing/managers.py", line 539, in _run_server
 [2021/11/12 01:05:16.124]     server = cls._Server(registry, address, authkey, serializer)
 [2021/11/12 01:05:16.124]   File "/opt/python/3.6/lib/python3.6/multiprocessing/managers.py", line 139, in __init__
 [2021/11/12 01:05:16.124]     self.listener = Listener(address=address, backlog=16)
 [2021/11/12 01:05:16.124]   File "/opt/python/3.6/lib/python3.6/multiprocessing/connection.py", line 438, in __init__
 [2021/11/12 01:05:16.124]     self._listener = SocketListener(address, family, backlog)
 [2021/11/12 01:05:16.124]   File "/opt/python/3.6/lib/python3.6/multiprocessing/connection.py", line 576, in __init__
 [2021/11/12 01:05:16.124]     self._socket.bind(address)
 [2021/11/12 01:05:16.124] OSError: AF_UNIX path too long
 [2021/11/12 01:05:16.125] Traceback (most recent call last):
 [2021/11/12 01:05:16.125]   File "kms_kmip_server.py", line 52, in <module>
 [2021/11/12 01:05:16.126]     main()
 [2021/11/12 01:05:16.126]   File "kms_kmip_server.py", line 46, in main
 [2021/11/12 01:05:16.126]     with server:
 [2021/11/12 01:05:16.126]   File "/data/mci/95a1a58fd6dec98b9c985856ff7214a8/drivers-tools/.evergreen/csfle/kmstlsvenv/lib/python3.6/site-packages/kmip/services/server/server.py", line 476, in __enter__
 [2021/11/12 01:05:16.126]     self.start()
 [2021/11/12 01:05:16.126]   File "/data/mci/95a1a58fd6dec98b9c985856ff7214a8/drivers-tools/.evergreen/csfle/kmstlsvenv/lib/python3.6/site-packages/kmip/services/server/server.py", line 243, in start
 [2021/11/12 01:05:16.126]     self.manager = multiprocessing.Manager()
 [2021/11/12 01:05:16.126]   File "/opt/python/3.6/lib/python3.6/multiprocessing/context.py", line 56, in Manager
 [2021/11/12 01:05:16.126]     m.start()
 [2021/11/12 01:05:16.126]   File "/opt/python/3.6/lib/python3.6/multiprocessing/managers.py", line 517, in start
 [2021/11/12 01:05:16.126]     self._address = reader.recv()
 [2021/11/12 01:05:16.126]   File "/opt/python/3.6/lib/python3.6/multiprocessing/connection.py", line 250, in recv
 [2021/11/12 01:05:16.127]     buf = self._recv_bytes()
 [2021/11/12 01:05:16.127]   File "/opt/python/3.6/lib/python3.6/multiprocessing/connection.py", line 407, in _recv_bytes
 [2021/11/12 01:05:16.127]     buf = self._recv(4)
 [2021/11/12 01:05:16.127]   File "/opt/python/3.6/lib/python3.6/multiprocessing/connection.py", line 383, in _recv
 [2021/11/12 01:05:17.339]     raise EOFError
 [2021/11/12 01:05:17.339] EOFError
```
https://spruce.mongodb.com/task/mongo_python_driver_tests_python_version_amazon1_test_encryption__platform~awslinux_auth_ssl~noauth_nossl_python_version~3.9_encryption~encryption_test_5.0_standalone_patch_370e1652ad97a2dbfd33400fc0e23a6c2fc4a5d5_618db7453e8e863fe33d8152_21_11_12_00_37_26/logs?execution=0&sortBy=STATUS&sortDir=ASC

Googling around I found this very helpful issue: https://github.com/broadinstitute/cromwell/issues/3647

TLDR: Setting $TMPDIR to a shorter path should resolve the issue.

**Comment by ShaneHarvey** [comment]
Debugging a few more failures all on Windows. Windows Python 3.10 times out after blocking for 27 minutes in `test_encryption.TestKmsTLSOptions.test_01_aws`
```
 [2021/11/12 20:30:23.823] test_explicit (test_encryption.TestGCPEncryption) ... ok
 [2021/11/12 20:57:28.670] Command stopped early: context canceled
 [2021/11/12 20:57:28.705] test_01_aws (test_encryption.TestKmsTLSOptions) ...
 [2021/11/12 20:57:28.705] Running task-timeout commands.
```
https://spruce.mongodb.com/task/mongo_python_driver_tests_windows_encryption__platform~windows_64_vsMulti_small_auth_ssl~noauth_nossl_python_version_windows~3.10_encryption~encryption_test_4.2_standalone_patch_370e1652ad97a2dbfd33400fc0e23a6c2fc4a5d5_618ec56fe3c3314c3170fc42_21_11_12_19_50_08/logs?execution=0&sortBy=STATUS&sortDir=ASC

Non-Python 3.6 tests on Windows fail with:
```
 [2021/11/12 20:28:00.459] FAIL: test_02_azure (test_encryption.TestKmsTLSOptions)
 [2021/11/12 20:28:00.459] ----------------------------------------------------------------------
 [2021/11/12 20:28:00.459] Traceback (most recent call last):
 [2021/11/12 20:28:00.459]   File "C:\data\mci\84ae93aa280a4192c655cbe26366f82a\src\pymongo\encryption.py", line 74, in _wrap_encryption_errors
 [2021/11/12 20:28:00.459]     yield
 [2021/11/12 20:28:00.459] ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host
 [2021/11/12 20:28:00.459] During handling of the above exception, another exception occurred:
 [2021/11/12 20:28:00.459] pymongo.errors.EncryptionError: [WinError 10054] An existing connection was forcibly closed by the remote host
 [2021/11/12 20:28:00.459] During handling of the above exception, another exception occurred:
 [2021/11/12 20:28:00.459] Traceback (most recent call last):
 [2021/11/12 20:28:00.459]   File "C:\data\mci\84ae93aa280a4192c655cbe26366f82a\src\test\test_encryption.py", line 1858, in test_02_azure
 [2021/11/12 20:28:00.459]     self.client_encryption_no_client_cert.create_data_key('azure', key)
 [2021/11/12 20:28:00.459] AssertionError: "certificate required|SSL handshake failed|EOF" does not match "[WinError 10054] An existing connection was forcibly closed by the remote host"
```
https://spruce.mongodb.com/task/mongo_python_driver_tests_windows_encryption__platform~windows_64_vsMulti_small_auth_ssl~noauth_nossl_python_version_windows~3.9_encryption~encryption_test_5.0_standalone_patch_370e1652ad97a2dbfd33400fc0e23a6c2fc4a5d5_618ec56fe3c3314c3170fc42_21_11_12_19_50_08/logs?execution=0&sortBy=STATUS&sortDir=ASC

There's also this error which might be related:
```
 [2021/11/12 20:24:48.813] dirname: missing operand
 [2021/11/12 20:24:48.813] Try 'dirname --help' for more information.
```

It would be great if we logged the stack thread of all the threads when we experience a EVG timeout. I think we can do that with a signal handler.

**Comment by ShaneHarvey** [comment]
Okay I think I've fixed the windows tests. The fix was to actually connect to the KMIP server before running the test suite. I'm not sure why this was required but it seems to fix the issue.

**Comment by ShaneHarvey** [comment]
Okay this is now ready for review.

**Review by juliusgeo — APPROVED** [review]
LGTM!

**Comment by ShaneHarvey** [comment]
I'm skipping the hanging windows 3.10 tests for now and I'll investigate in a new ticket.

File: pymongo/encryption_options.py [hunk 1]
```diff
@@ -55,9 +55,8 @@ def __init__(self, kms_providers, key_vault_namespace,
         See :ref:`automatic-client-side-encryption` for an example.
 
         :Parameters:
-          - `kms_providers`: Map of KMS provider options. Two KMS providers
-            are supported: "aws" and "local". The kmsProviders map values
-            differ by provider:
+          - `kms_providers`: Map of KMS provider options.The kmsProviders map
```
Comment (behackett) [inline pymongo/encryption_options.py:58]: Missing space before the second sentence?

File: test/test_encryption.py [hunk 2]
```diff
@@ -679,6 +693,10 @@ class TestDataKeyDoubleEncryption(EncryptionIntegrationTest):
             'location': 'global',
             'keyRing': 'key-ring-csfle',
             'keyName': 'key-name-csfle'},
+        'kmip': {
+            'keyId': '1',
+            'endpoint': KMIP['endpoint'],  # TODO: is this needed?
```
Comment (behackett) [inline test/test_encryption.py:698]: Well, is it?

File: pymongo/encryption_options.py (hunk 1 above)
Comment (ShaneHarvey) [reply pymongo/encryption_options.py]: Done.

File: test/test_encryption.py (hunk 2 above)
Comment (ShaneHarvey) [reply test/test_encryption.py]: Good catch, the spec actually says to use {} as the kmip key: https://github.com/mongodb/specifications/tree/master/source/client-side-encryption/tests#data-key-and-double-encryption

Fixed.

**Review by behackett — APPROVED** [review]
